doc/designs/ech-api.md

```diff
@@ -0,0 +1,590 @@
+ascii hex or binary) each of which may suit different applications.  ECHConfig
+values may also be provided embedded in the DNS wire encoding of HTTPS or SVCB
+resource records or in the equivalent zone file presentation format.
+
+``OSSL_ech_find_echconfigs()`` attempts to find and return the (possibly empty)
+set of ECHConfig values from a buffer containing one of the encoded forms
+described above. Each successfully returned ECHConfigList will have
+exactly one ECHConfig, i.e., a single public value.
+
+```c
+int OSSL_ech_find_echconfigs(int *num_echs,
+                             unsigned char ***echconfigs, size_t **echlens,
+                             const unsigned char *val, size_t len);
+```
+
+``OSSL_ech_find_echconfigs()`` returns the number of ECHConfig values from the
+input (``val``/``len``) successfully decoded  in the ``num_echs`` output.  If
+no ECHConfig values values are encountered (which can happen for good HTTPS RR
+values) then ``num_echs`` will be zero but the function returns 1. If the
+input contains more than one (syntactically correct) ECHConfig, then only
+those that contain locally supported options (e.g. AEAD ciphers) will be
+returned. If no ECHConfig found has supported options then none will be
+returned and the function will return 0.
+
+After a call to ``OSSL_ech_find_echconfigs()``, the application can make a
+sequence of calls to ``SSL_ech_set1_echconfig()`` for each of the ECHConfig
+values found.  (The various output buffers must be freed by the client
+afterwards, see the example code in
+[``test/ech_test.c``](https://github.com/sftcd/openssl/blob/ECH-draft-13c/test/ech_test.c).)
+
+Clients can additionally more directly control the values to be used for inner
+and outer SNI and ALPN values via specific APIs. This allows a client to
+override the ``public_name`` present in an ECHConfigList that will otherwise
+be used for the outer SNI. The ``no_outer`` input allows a client to emit an
+outer CH with no SNI at all.
+
+```c
+int SSL_ech_set_server_names(SSL *s, const char *inner_name,
+                             const char *outer_name, int no_outer);
+int SSL_ech_set_outer_server_name(SSL *s, const char *outer_name, int no_outer);
+int SSL_ech_set_outer_alpn_protos(SSL *s, const unsigned char *protos,
+                                  unsigned int protos_len);
+int SSL_CTX_ech_set_outer_alpn_protos(SSL_CTX *s, const unsigned char *protos,
+                                      unsigned int protos_len);
+```
+
+If a client attempts ECH but that fails, or sends an ECH-GREASEd CH, to
+an ECH-supporting server, then that server may return an ECH "retry-config"
+value that the client could choose to use in a subsequent connection. The
+client can detect this situation via the ``SSL_ech_get_status()`` API and
+can access the retry config value via:
+
+```c
+int SSL_ech_get_retry_config(SSL *s, unsigned char **ec, size_t *eclen);
```

Comment:
Is this binary buffer another ECHConfig?

Reply:
It's an ECHConfigList again. So if a client wanted to do a retry (with another TLS connection/socket) it could feed that buffer into the next call to ``SSL_ech_set1_echconfig()`.

Reply:
But previously we were talking about ECHConfigs not ECHConfigLists. If we had type safety through a proper object this would be clear and self-documenting.

Reply:
Is it better to get this as an `OSSL_ECHSTORE`?